### PR TITLE
wave3: Sparse Mixture-of-Experts FFN Layers (cross-dataset)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Replacing the standard FFN layers with sparse Mixture-of-Experts (MoE) FFN increases effective model capacity without proportionally increasing compute. With n_experts=8 and top-k=2, each token dynamically routes through the 2 most relevant experts, enabling specialization across different geometric regions (boundary layers, wake regions, leading/trailing edges). This should improve accuracy on all four datasets.

## Implementation

Add to `train.py`:

```python
class SparseMoEFFN(nn.Module):
    """Sparse Mixture-of-Experts FFN with top-k routing."""
    def __init__(self, dim, n_experts=8, top_k=2, hidden_dim=None, load_balance_weight=0.01):
        super().__init__()
        hidden_dim = hidden_dim or dim * 4
        self.n_experts = n_experts
        self.top_k = top_k
        self.load_balance_weight = load_balance_weight
        
        # Router
        self.router = nn.Linear(dim, n_experts, bias=False)
        
        # Expert FFNs
        self.experts = nn.ModuleList([
            nn.Sequential(
                nn.Linear(dim, hidden_dim),
                nn.GELU(),
                nn.Linear(hidden_dim, dim),
            )
            for _ in range(n_experts)
        ])
        self.load_balance_loss = 0.0
    
    def forward(self, x):
        B, N, D = x.shape
        x_flat = x.view(-1, D)  # [B*N, D]
        
        # Route
        logits = self.router(x_flat)  # [B*N, E]
        probs = torch.softmax(logits, dim=-1)
        top_k_probs, top_k_indices = torch.topk(probs, self.top_k, dim=-1)
        top_k_probs = top_k_probs / top_k_probs.sum(dim=-1, keepdim=True)
        
        # Load balancing loss (auxiliary)
        if self.training and self.load_balance_weight > 0:
            f = probs.mean(0)  # fraction of tokens per expert
            P = probs.mean(0)
            self.load_balance_loss = self.load_balance_weight * self.n_experts * (f * P).sum()
        
        # Dispatch and combine
        output = torch.zeros_like(x_flat)
        for i in range(self.top_k):
            expert_idx = top_k_indices[:, i]
            expert_w = top_k_probs[:, i:i+1]
            for e in range(self.n_experts):
                mask = (expert_idx == e)
                if mask.any():
                    output[mask] += expert_w[mask] * self.experts[e](x_flat[mask])
        
        return output.view(B, N, D)
```

Add CLI arguments:
```python
parser.add_argument('--moe-layers', type=str, default='',
    help='Comma-separated layer indices to replace with MoE FFN (e.g. "1,2,3")')
parser.add_argument('--moe-n-experts', type=int, default=8)
parser.add_argument('--moe-load-balance-weight', type=float, default=0.01)
```

Add MoE load balance loss to total loss:
```python
if args.moe_layers:
    for layer in model.transformer_layers:
        if hasattr(layer.ffn, 'load_balance_loss'):
            total_loss = total_loss + layer.ffn.load_balance_loss
```

Use `--wandb_group wave3-moe-ffn-layers` for grouping.

### TandemFoil
```bash
python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --moe-layers 1,2,3 --moe-n-experts 8 --moe-load-balance-weight 0.01 \
  --epochs 999 --wandb_group wave3-moe-ffn-layers
```

### TandemFoil Paper
```bash
python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --moe-layers 1,2,3 --moe-n-experts 8 --moe-load-balance-weight 0.01 \
  --epochs 999 --wandb_group wave3-moe-ffn-layers
```

### AirfRANS
```bash
python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --moe-layers 1,2 --moe-n-experts 8 --moe-load-balance-weight 0.01 \
  --epochs 999 --wandb_group wave3-moe-ffn-layers
```

### DrivAerML
```bash
SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --moe-layers 1,2,3,4 --moe-n-experts 8 --moe-load-balance-weight 0.01 \
  --wandb_group wave3-moe-ffn-layers
```

## Current Baselines

| Dataset | Metric | Best Value | PR |
|---------|--------|-----------|-----|
| TandemFoil | val_primary/surface_pressure_mae | 26.06 | #2887 |
| TandemFoil Paper | val_primary/field_mse | TBD | #2947/2948/2949 |
| AirfRANS | val_primary/surface_mse | 0.000627 | #2902 |
| DrivAerML | val_primary/surface_rel_l2_pct | 3.997% | #2898 |

## Expected Outcome

MoE FFN allows the model to specialize different experts for different flow regimes (attached flow, separated wake, stagnation regions) without requiring explicit labels. The sparse routing (top-k=2 out of 8) keeps compute comparable to a single expert while quadrupling effective expert capacity. Expect 2-5% improvement across datasets, with larger gains on DrivAerML due to greater geometric complexity.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shoya:

## Results

### Summary

All-layer Sparse MoE-FFN (8 experts, top-k=2) tested across 3 datasets with `--moe-layers` applying MoE to ALL Transolver blocks. **MoE does not beat the dense baseline on any dataset.** DrivAerML suffers catastrophic recurring expert collapse at cosine LR restarts. TandemFoil Paper was skipped (no data files available).

### DrivAerML — MoE layers 1,2,3,4 (all 4 blocks)

| Metric | Baseline | MoE Best (ep34) | MoE Final (ep48) |
|--------|----------|-----------------|-------------------|
| val_primary/surface_rel_l2_pct | **3.997%** | 18.11% (4.5x worse) | 54.63% (13.7x worse) |

- **W&B run**: `hlfpdbd6`
- **Epochs completed**: 48 (still running, but terminal collapse)
- **Expert collapse timeline**:
  - ep1–35: Functional but never competitive. Best 18.11% at ep34 (stable phase, lb_loss ~4.0)
  - ep36: **First collapse** — loss 0.468→0.112, grad_norm 5.4, lb_loss 8.6, val 60.0%
  - ep37: Expert 6 at 50.0%, expert 7 DEAD (0.0%), val 38.4%
  - ep38: Expert 6 at 49.2%, expert 7 still dead, val 56.3%
  - ep39–45: Partial recovery, val oscillating 20–32%
  - ep46: **Second collapse** — loss 0.56, grad_norm 8.7, lb_loss 10.1, val 54.5%
  - ep47: Loss 0.80, grad_norm 10.3, lb_loss 14.7, val 63.1% (worst)
  - ep48: **Terminal** — experts 1,4 DEAD (0.0%), experts 0,2 each at ~48%, val 54.6%
- **Key finding**: Every cosine LR restart (T_max=30) triggers expert collapse. The 0.01 load-balance weight is insufficient to prevent the router from collapsing when learning rate spikes. This is identical to PR #2935's finding — all-layer MoE makes it *worse* (collapse occurs earlier and more severely with 4 MoE layers vs 2-4 in PR #2935).

### AirfRANS — MoE layers 1,2 (all 2 blocks)

| Metric | Baseline | MoE Best (ep32) | MoE Latest (ep37) |
|--------|----------|-----------------|-------------------|
| val_primary/surface_mse | **0.000627** | 0.00987 (15.7x worse) | 0.0623 |

- **W&B run**: `255hczs3`
- **Epochs completed**: 37 (still running)
- **Routing**: Stable throughout. No expert collapse. Load-balance loss steady at ~2.02. Expert fractions in 10–16% range (healthy for 8 experts).
- **Val trajectory**: 0.767→0.131→0.055→0.020→0.00987 (improving at each cosine trough, ep5/15/25/32)
- **Key finding**: MoE routing is stable on AirfRANS (only 2 MoE layers, smaller model), but the model converges 15.7x slower than the dense baseline. The extra capacity from 8 experts is not useful — AirfRANS has only 2 Transolver blocks, so the dense MLP is already sufficient.

### TandemFoil — MoE layers 1,2,3 (all 3 blocks)

| Metric | Baseline (EMA) | Baseline (no-EMA) | MoE Best (ep13) | MoE Latest (ep14) |
|--------|---------------|-------------------|-----------------|-------------------|
| val_primary/surface_pressure_mae | **26.06** | ~30.10 | 51.27 (1.7x worse) | 58.93 |

- **W&B run**: `idelnafw`
- **Epochs completed**: 14 (still running, early)
- **Routing**: Healthy throughout. Layer 0 has mild expert preference (expert 6 at ~17%), but no collapse. Load-balance loss ~3.1.
- **Val trajectory**: 173→118→98→83→65→51 (steadily improving at cosine troughs)
- **Key finding**: TandemFoil is still early (14 epochs, cosine T_max=10) and may continue improving, but the trajectory suggests convergence well above the 26.06 baseline. The MoE variant uses ~8x more FFN parameters but shows no benefit over the dense baseline.

### TandemFoil Paper — SKIPPED

No data files available at the expected path. Same as PR #2935.

### Commands Used

```bash
# DrivAerML (GPU 2-7)
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2,3,4,5,6,7 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --moe-layers 1,2,3,4 --moe-n-experts 8 --moe-load-balance-weight 0.01 --wandb-group wave3-moe-ffn-layers --wandb-name shoya/drivaerml-moe-all-layers --agent shoya

# AirfRANS (GPU 1)
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --moe-layers 1,2 --moe-n-experts 8 --moe-load-balance-weight 0.01 --epochs 999 --wandb-group wave3-moe-ffn-layers --wandb-name shoya/airfrans-moe-all-layers --agent shoya

# TandemFoil (GPU 0)
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --moe-layers 1,2,3 --moe-n-experts 8 --moe-load-balance-weight 0.01 --epochs 999 --wandb-group wave3-moe-ffn-layers --wandb-name shoya/tandemfoil-moe-all-layers --agent shoya
```

### Peak Memory

| Dataset | GPU | VRAM Used |
|---------|-----|-----------|
| TandemFoil | GPU 0 | ~81 GB |
| AirfRANS | GPU 1 | ~95 GB |
| DrivAerML | GPU 2 (lead) | ~19 GB |

### What Happened

**MoE-FFN fails across all three datasets**, confirming and extending PR #2935's findings:

1. **DrivAerML**: Catastrophic expert collapse is triggered by cosine LR restarts. When the LR spikes at cycle boundaries (ep36, ep46), the router destabilizes — dominant experts absorb all routing mass while others die. The 0.01 load-balance penalty is insufficient. All-layer MoE (4 layers) collapses *more severely* than last-N MoE from PR #2935, as the instability propagates through more layers.

2. **AirfRANS**: No routing collapse (only 2 layers), but converges 15.7x slower than baseline. The sparse routing adds complexity without benefit for this small model.

3. **TandemFoil**: Healthy routing but 1.7x worse even at this early stage. The additional capacity from 8 experts per layer doesn't improve prediction quality.

The hypothesis that "MoE FFN allows the model to specialize different experts for different flow regimes" is not supported. The router does not learn physically meaningful specialization — it either collapses (DrivAerML) or converges slower (AirfRANS, TandemFoil). The dense FFN is a better inductive bias for these CFD tasks.

### Suggested Follow-ups

1. **Router-freezing during LR warmup**: Freeze router weights during the first N steps of each cosine cycle to prevent collapse at LR restarts. This addresses the proximate cause but not the fundamental issue.
2. **Expert-choice routing** (Zhou et al., 2022): Instead of token-choice top-k, let each expert select its top tokens. This guarantees load balance by construction, eliminating the collapse mechanism entirely.
3. **Conditional computation without routing**: Instead of MoE, try simple depth-wise conditional computation (e.g., skip certain layers for easy tokens). This gets compute savings without the routing instability.
4. **Abandon MoE-FFN for these tasks**: Two PRs (#2935, #2960) now show MoE-FFN consistently fails. The CFD operator learning task may not have the token-level diversity that MoE was designed for (unlike NLP where tokens have vastly different semantics). Recommend focusing on other architectural improvements instead.

**Comment by tcapelle** [comment]
ADVISOR REVIEW — PR #2960 wave3: Sparse Mixture-of-Experts FFN Layers

**W&B Verified Metrics (independent check):**
| Dataset | Run ID | Best Val Metric | Epoch | Final State |
|---|---|---|---|---|
| DrivAerML | hlfpdbd6 | 18.1122% | ~33 | Collapsed by ep49 |
| AirfRANS | 255hczs3 | 0.00986833 | ~31 | Running (stalled) |
| TandemFoil | idelnafw | 51.2683 | ~12 | Running (stalled) |

**Baselines to beat:**
- DrivAerML: 3.997% (PR #2898)
- TandemFoil: 26.06 (PR #2887)
- AirfRANS: 0.000598 (PR #2906)

**Verdict: Close — strong negative across all three datasets. Abandon MoE-FFN line.**

Every dataset is dramatically worse than baseline:
- DrivAerML: 18.11% vs 3.997% baseline (4.5x worse), then terminal collapse at ep48
- AirfRANS: 0.00987 vs 0.000598 baseline (16.5x worse)
- TandemFoil: 51.27 vs 26.06 baseline (nearly 2x worse)

**Root Cause Analysis:**

This confirms and extends the finding from PR #2935 (which already established MoE failure on DrivAerML). Two independent experiments now establish the same failure pattern across all three CFD datasets.

The failure mechanism is clear: cosine LR restarts create sudden LR spikes that destabilize the expert router. At a spike, the router receives a large gradient update and collapses all tokens to a single expert — expert collapse. The load-balancing weight of 0.01 is insufficient to prevent this; the LR spike signal overwhelms the lb_weight regularization.

CFD operator learning tasks lack the token-level semantic diversity that MoE was designed for. In NLP, tokens carry distinct semantic roles (noun, verb, clause, etc.) that provide natural diversity for experts to specialize on. In CFD mesh node features, all nodes are physical field values at different spatial locations — there is much less discrete semantic structure for experts to exploit. The routing mechanism consequently has nothing meaningful to latch onto, and any perturbation (LR spike, batch noise) collapses it.

**Conclusion:**

MoE-FFN is not a viable direction for this problem family. Two experiments confirm it. GPU time is better spent on other hypotheses. We are closing this line.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2960 because ADVISOR: Closing as strong negative. MoE-FFN fails catastrophically on all three CFD datasets (DrivAerML 18.11% vs 3.997% baseline, AirfRANS 0.00987 vs 0.000598, TandemFoil 51.27 vs 26.06). Expert collapse triggered by cosine LR restart spikes — lb_weight=0.01 insufficient. This confirms PR #2935 findings and extends them across all datasets. Abandoning MoE-FFN line. Full review in PR comments..
